clang/lib/Sema/AnalysisBasedWarnings.cpp

```diff
@@ -696,12 +661,12 @@ static void CheckFallThroughForBody(Sema &S, const Decl *D, const Stmt *Body,
   if (CD.checkDiagnostics(Diags, ReturnsVoid, HasNoReturn))
       return;
   SourceLocation LBrace = Body->getBeginLoc(), RBrace = Body->getEndLoc();
-  auto EmitDiag = [&](SourceLocation Loc, unsigned DiagID) {
+  auto EmitDiag = [&](SourceLocation Loc, unsigned DiagID, unsigned FunMode) {
```

Comment:
This lambda has a capture-default of `&` anyway, so just use `CD.FunMode` in the lambda imo.

Reply:
Hmm, looking at how this has turned out, I think the entire lambda body can just be
```c++
    if (DiagID)
      S.Diag(Loc, DiagID) << CD.FunMode;
```
and then use `EmitDiag` instead of `S.Diag` everywhere in the `switch` statement below. That way we don’t have to think too hard about when a diag id might or might not be `0`. That also should allow us to remove the `IsCoroutine`-related checks again.

Reply:
Unfortunately, in `diag_FallThrough_HasNoReturn` case, we have one argument, and in `diag_FallThrough_ReturnsNonVoid` we have two arguments, so I just removed `EmitDiag` since it can not check how many arguments diag message have and use a proper branch.

Reply:
Ah, that’s annoying. We could make the lambda variadic and use a fold expression, but it’s not really worth it for a grand total of 2 function calls, so this is fine